src/v/cluster/decommissioning_monitor.cc

```diff
@@ -0,0 +1,222 @@
+#include "cluster/decommissioning_monitor.h"
+
+#include "cluster/logger.h"
+#include "cluster/members_table.h"
+#include "cluster/partition_allocator.h"
+#include "cluster/topic_table.h"
+#include "cluster/topics_frontend.h"
+#include "cluster/types.h"
+#include "config/configuration.h"
+#include "model/metadata.h"
+#include "model/timeout_clock.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/util/later.hh>
+
+#include <absl/container/flat_hash_set.h>
+
+#include <algorithm>
+#include <exception>
+#include <iterator>
+
+namespace cluster {
+
+decommissioning_monitor::decommissioning_monitor(
+  ss::sharded<cluster::topics_frontend>& fr,
+  ss::sharded<topic_table>& tt,
+  ss::sharded<partition_allocator>& pal,
+  consensus_ptr raft0,
+  ss::sharded<ss::abort_source>& as)
+  : _topics_frontend(fr)
+  , _topics(tt)
+  , _allocator(pal)
+  , _raft0(raft0)
+  , _as(as)
+  , _retry_timeout(config::shard_local_cfg().decomission_monitor_retry_ms()) {
+    _retry_timer.set_callback([this] { dispatch_decommissioning(); });
+}
+
+ss::future<> decommissioning_monitor::stop() {
+    _retry_timer.cancel();
+    return _bg.close();
+}
+
+void decommissioning_monitor::start() { _retry_timer.arm(_retry_timeout); }
+
+void decommissioning_monitor::dispatch_decommissioning() {
+    (void)ss::with_gate(_bg, [this] {
+        return try_decommission().finally([this] {
+            if (!_bg.is_closed()) {
+                _retry_timer.arm(_retry_timeout);
+            }
+        });
+    }).handle_exception([](const std::exception_ptr& e) {
+        // just log an exception, will retry
+        vlog(
+          clusterlog.trace,
+          "error encountered while handling nodes decomissioning - {}",
+          e);
+    });
+}
+
+ss::future<> decommissioning_monitor::try_decommission() {
+    // nothing to do, just return
+    if (_decommissioned.empty() && _reallocations.empty()) {
+        co_return;
+    }
+
+    co_await ss::parallel_for_each(
+      _reallocations,
+      [this](reallocation_meta& meta) { return reallocate_replica_set(meta); });
+
+    // remove finished meta
+    std::erase_if(_reallocations, [](const reallocation_meta& meta) {
+        return meta.is_finished;
+    });
+
+    auto to_remove = co_await _allocator.invoke_on(
+      partition_allocator::shard, [](partition_allocator& allocator) {
+          std::vector<model::node_id> ids;
+          for (const auto& [id, node] : allocator.allocation_nodes()) {
+              if (node->is_decommissioned() && node->empty()) {
+                  ids.push_back(id);
+              }
+          }
+          return ids;
+      });
+
+    // no nodes to reamove from cluster
+    if (to_remove.empty() || !_raft0->is_leader()) {
```

Comment:
Maybe the is_leader check can be added to the top of the method, performed earlier?